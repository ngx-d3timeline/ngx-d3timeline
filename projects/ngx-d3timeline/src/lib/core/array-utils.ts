export function sum(...args: number[]) {
  return args.reduce((a, b) => a + b);
}
