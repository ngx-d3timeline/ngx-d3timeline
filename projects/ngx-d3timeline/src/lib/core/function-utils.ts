export function partialApply<TInput, TAppliedArgs extends any[], TOut>(
  func: (input: TInput, ...appliedArgs: TAppliedArgs) => TOut
): (...appliedArgs: TAppliedArgs) => (input: TInput) => TOut {
  return (...appliedArgs: TAppliedArgs) => (input: TInput) =>
    func(input, ...appliedArgs);
}

export function pipe<T>(
  ...args: ((...args: any[]) => T)[]
): (...t: any[]) => T {
  return x => args.reduce((acc, curr) => curr(acc), x);
}

export function clampZero(x: number) {
  return Math.max(x, 0);
}
