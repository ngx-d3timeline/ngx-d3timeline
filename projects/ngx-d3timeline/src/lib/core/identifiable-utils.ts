import { Identifiable, identifier } from './types';

export function findIdentifiable<T extends Identifiable>(
  identifiables: T[],
  id: identifier
): T {
  return id && identifiables.find(identifiable => identifiable.id === id);
}
