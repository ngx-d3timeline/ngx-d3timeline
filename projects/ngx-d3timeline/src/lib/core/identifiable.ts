export type Identifier = string | number;

export interface Identifiable {
  id: Identifier;
}

export function findIdentifiable<T extends Identifiable>(
  identifiables: T[],
  id: Identifier
): T {
  return id && identifiables.find(identifiable => identifiable.id === id);
}
