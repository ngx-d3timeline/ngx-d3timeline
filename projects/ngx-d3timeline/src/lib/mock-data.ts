import { TimelineEvent } from './timeline-event';

export const mockData: TimelineEvent[] = [
  {
    id: 1,
    start: new Date('2017-09-23T08:45:00'),
    finish: new Date('2017-09-23T11:47:38'),
    type: 'Driving',
    series: 'Series 1'
  },
  {
    id: 2,
    start: new Date('2017-09-22T04:09:21'),
    finish: new Date('2017-09-22T08:39:21'),
    type: 'Driving',
    series: 'Series 1'
  },
  {
    id: 3,
    start: new Date('2017-09-22T08:39:21'),
    finish: new Date('2017-09-22T09:24:21'),
    type: 'DriveBreak',
    series: 'Series 1',
    sequence: 1
  },
  {
    id: 13,
    start: new Date('2017-09-22T10:00:00'),
    finish: new Date('2017-09-22T10:08:31'),
    type: 'Driving',
    series: 'Series 2'
  },
  {
    id: 14,
    start: new Date('2017-09-22T09:15:00'),
    finish: new Date('2017-09-22T09:16:16'),
    type: 'Waiting',
    series: 'Series 2',
    sequence: 6
  },
  {
    id: 15,
    start: new Date('2017-09-22T13:45:00'),
    finish: new Date('2017-09-22T14:52:27'),
    type: 'Driving',
    series: 'Series 2'
  },
  {
    id: 41,
    start: new Date('2017-09-23T08:45:00'),
    finish: new Date('2017-09-23T11:47:38'),
    type: 'Driving',
    series: 'Series 3'
  },
  {
    id: 42,
    start: new Date('2017-09-22T04:09:21'),
    finish: new Date('2017-09-22T08:39:21'),
    type: 'Driving',
    series: 'Series 3'
  },
  {
    id: 43,
    start: new Date('2017-09-22T08:39:21'),
    finish: new Date('2017-09-22T09:24:21'),
    type: 'DriveBreak',
    series: 'Series 3',
    sequence: 1
  }
];
