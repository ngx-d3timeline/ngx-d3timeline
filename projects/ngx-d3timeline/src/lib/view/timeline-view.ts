export class TimelineView {
  readonly width: number;
  readonly height: number;
  readonly margin: number;

  readonly bottom: number;
  readonly left: number;
  readonly right: number;
  readonly top: number;

  constructor({
    width,
    height,
    margin
  }: {
    width: number;
    height: number;
    margin: number;
  }) {
    this.width = width;
    this.height = height;
    this.margin = margin;

    this.bottom = height - margin;
    this.left = 0;
    this.right = width - margin;
    this.top = 0;
  }
}
