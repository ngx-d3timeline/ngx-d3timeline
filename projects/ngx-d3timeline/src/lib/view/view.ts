export class View {
  readonly width: number;
  readonly height: number;
  readonly margin = 50;

  readonly left = this.margin;
  readonly top = this.margin;
  readonly right: number;
  readonly bottom: number;

  constructor([width, height]: [number, number]) {
    this.width = width;
    this.height = height;

    this.bottom = height - this.margin;
    this.right = width - this.margin;
  }
}
