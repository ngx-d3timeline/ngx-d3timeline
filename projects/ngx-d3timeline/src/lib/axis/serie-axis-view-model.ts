import { ScaleBand, scaleBand } from 'd3-scale';
import { TimelineEvent } from '../timeline-event';
import { View } from '../view/view';

export class SeriesAxisViewModel {
  private readonly bandscale: ScaleBand<string>;

  constructor(data: TimelineEvent[], view: View) {
    this.bandscale = this.createBandscale(data, view);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.bandscale.domain().map(tick => ({
      label: tick,
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.bandscale.range()[1];
  }

  get bandwidth() {
    return this.bandscale.bandwidth();
  }

  private createBandscale(data: TimelineEvent[], view: View) {
    return scaleBand()
      .domain([...new Set(data.map(d => d.series))])
      .range([0, view.width - view.margin]);
  }

  private tickTransform(tick: string) {
    return `translate(${this.bandscale(tick) +
      this.bandscale.bandwidth() / 2}, 0)`;
  }
}
