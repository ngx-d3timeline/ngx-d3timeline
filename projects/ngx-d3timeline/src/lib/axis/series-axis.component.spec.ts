import { TestBed, ComponentFixture } from '@angular/core/testing';
import { SeriesAxisComponent } from './series-axis.component';
import { SeriesAxisViewModel } from './serie-axis-view-model';
import { mockData } from '../mock-data';
import { TimelineView } from '../view/timeline-view';

describe('SeriesAxisComponent', () => {
  let fixture: ComponentFixture<SeriesAxisComponent>;
  const timelineView: TimelineView = new TimelineView(800, 600, 100);

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [SeriesAxisComponent]
    });

    fixture = TestBed.createComponent(SeriesAxisComponent);
  });

  it('should not render if view model is null', () => {
    const vm = null;
    fixture.componentInstance.vm = vm;

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const vm = new SeriesAxisViewModel(mockData, timelineView);
    fixture.componentInstance.vm = vm;

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
