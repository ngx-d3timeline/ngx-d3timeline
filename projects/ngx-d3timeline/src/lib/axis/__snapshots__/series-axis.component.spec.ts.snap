// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeriesAxisComponent should not render if view model is null 1`] = `
<div
  id="root0"
  ngx-d3timeline-series-axis=""
>
  
</div>
`;

exports[`SeriesAxisComponent should render correctly 1`] = `
<div
  id="root1"
  ngx-d3timeline-series-axis=""
>
  
  <g
    class="series-axis-group"
  >
    
    <text
      dy="-2"
      transform="translate(116.66666666666667, 0)"
    >
       Series 1 
    </text>
    <text
      dy="-2"
      transform="translate(350, 0)"
    >
       Series 2 
    </text>
    <text
      dy="-2"
      transform="translate(583.3333333333334, 0)"
    >
       Series 3 
    </text>
    <line
      class="series-axis-line"
      x2="700"
    />
  </g>
</div>
`;
